app/assets/src/components/ui/containers/ContextPlaceholder.jsx

```diff
@@ -0,0 +1,140 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+export default class ContextPlaceholder extends React.PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      style: this.getStyle(),
+      open: this.props.open
+    };
+  }
+
+  componentDidMount() {
+    if (this.props.closeOnOutsideClick)
+      document.addEventListener("mousedown", this.handleOutClick);
+  }
+
+  componentWillUnmount() {
+    if (this.props.closeOnOutsideClick)
+      document.removeEventListener("mousedown", this.handleOutClick);
+  }
+
+  handleOutClick = () => {
+    if (this.placeholderRef && !this.placeholderRef.contains(event.target)) {
+      this.setState({ open: false });
+      this.props.onClose && this.props.onClose();
+    }
+  };
+
+  handlePlaceholderRef = placeholder => {
+    this.placeholderRef = placeholder;
+    this.setState({ style: this.getStyle() });
+  };
+
+  getStyle() {
+    let style = { position: "absolute" };
+    if (this.props.context && this.placeholderRef) {
+      Object.assign(style, this.getPosition());
+    }
+    return style;
+  }
+
+  getPosition() {
+    const contextRect = this.props.context.getBoundingClientRect();
```

Comment:
Could you move this logic to a pure utility function in a separate file?

Reply:
The positioning logic here is quite complicated, and I'm not sure it will work for all cases (this stuff tends to be finicky). You might need to use a `<Portal>` to move `<ContextPlaceholder>` elements higher on the DOM tree so it's not the child of any elements that are `position: absolute` or `position: relative`. 

No change needed, but just a note that I think we should try to avoid writing this kind of complicated logic without a good reason. Awesome that it works though!

Reply:
this logic gets element position but also sets the context. I do not think that is that generic.

Reply:
For my first comment (moving into pure utility function), I meant that since `getPosition` is taking in parameters (such as contextRect and placeholderRect) and outputting a value (the position), it would be easy to factor it out into a separate file like `ContextPlaceholder/utils.js` so that all the business logic isn't cluttering up the file. This file is pretty short though, so not a big deal.

Reply:
It takes at least three more parameters: `horizontalOffset`, `verticalOffset`, `position`, which makes it a bit more specific to this element. 
I also not sure if you ever should use it outside of this element, or use this element instead.